From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Revanth Kumar Uppala <ruppala@nvidia.com>
Date: Tue, 20 Sep 2022 12:33:12 +0530
Subject: [PATCH 034/243] NVIDIA: SAUCE: net: phy: aqr113c: Read WOL status
 register.

Issue: In case of WOL disable, WOL status
is being shown as 'g' instead of 'd'.

Fix: Read proper WOL status register to set
status of WOL in ethtool output.

Bug 3789538

Change-Id: Ic1949c2ef71d5ad7d068ecef458464c2137ee1ee
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2909751
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index 96aee6a22b5f..f1f1a9d63246 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -56,6 +56,8 @@
 #define MDIO_PHYXS_VEND_IF_STATUS_TYPE_OCSGMII	10
 #define MDIO_PHYXS_VEND_IF_STATUS_TX_READY	BIT(12)
 
+#define MDIO_AN_RSVD_VEND_STATUS3		0xc812
+
 #define MDIO_AN_VEND_PROV			0xc400
 #define MDIO_AN_VEND_PROV_1000BASET_FULL	BIT(15)
 #define MDIO_AN_VEND_PROV_1000BASET_HALF	BIT(14)
@@ -1461,13 +1463,13 @@ static void aqr113c_get_wol(struct phy_device *phydev, struct ethtool_wolinfo *w
 {
 	u16 val;
 
-	val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL7);
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RSVD_VEND_STATUS3);
 	if (val < 0)
 		return;
 
+	wol->supported = WAKE_MAGIC;
 	if (val & 0x1) {
 		wol->wolopts = WAKE_MAGIC;
-		wol->supported = WAKE_MAGIC;
 	}
 }
 
-- 
2.43.0

